drivers/include/tla2528.hpp

```diff
@@ -0,0 +1,135 @@
+
+  tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address = default_address);
+  /**
+   * @brief set what service a pin will provide
+   */
+  void set_pin_mode(pin_mode p_mode, hal::byte p_channel);
+
+  // If channel is not set to correct config then undefined behavior will occur
+  /**
+   * @brief set output level pin
+   *
+   * If the pin is not an output pin the level will be used once it changes to
+   * an output pin.
+   */
+  void set_digital_out(hal::byte p_channel, bool level);
+  /**
+   * @brief set output levels on pins
+   *
+   * @param p_value The byte is used as a bit feild of bool values. i.e the 0th
+   * bit in the byte will set the 0 pin.
+   *
+   * If the pin is not an output pin the level will be used once it changes to
+   * an output pin output.
+   */
+  void set_digital_out(hal::byte p_values);
+  /**
+   * @brief read output state of an output pin
+   *
+   * If the pin is not an output pin the returned state will be used once it
+   * changes to an output pin.
+   */
+  bool get_digital_out(hal::byte p_channel);
+  /**
+   * @brief read the level of an pin.
+   */
+  bool get_digital_in(hal::byte p_channel);
+  /**
+   * @brief read the level of a pin
+   *
+   * @return The byte is used as a bit feild of bool values. i.e the 0th bit in
+   * the byte will be the 0 pin's level.
+   */
+  hal::byte get_digital_in();
+
+  /**
+   * @brief read the adc output of a pin.
+   */
+  float get_analog_in(hal::byte p_channel);
+
+  friend tla2528_adc;
+  friend tla2528_input_pin;
+  friend tla2528_output_pin;
+
+private:
+  // i2c address for no resistors
+  static constexpr hal::byte default_address = 0x10;
+
+  enum op_codes : hal::byte
+  {
+    single_register_read = 0b0001'0000,
+    single_register_write = 0b0000'1000,
+    set_bit = 0b0001'1000,
+    clear_bit = 0b0010'0000,
+    // Continuously reads data from a group of registers. Provide the first
+    // address to read from, if it runs out of valid addresses to read, it
+    // returns zeros. (See Figure 30 on datasheet)
+    continuous_register_read = 0b0011'0000,
+    // Continuously writes data to a group of registers. Provide the first
+    // address to write to.The data sent will automatically write the data to
+    // the next register in ascending order. (See Figure 32 on datasheet)
+    continuous_register_write = 0b0010'1000
+  };
+
+  enum register_addresses : hal::byte
+  {
+    system_status = 0x0,
+    general_cfg = 0x1,
+    data_cfg = 0x2,
+    osr_cfg = 0x3,
+    opmode_cfg = 0x4,
+    pin_cfg = 0x5,
+    gpio_cfg = 0x7,
+    gpo_drive_cfg = 0x9,
+    gpo_value = 0xB,
+    gpi_value = 0xD,
+    sequence_cfg = 0x10,
+    channel_sel = 0x11,
+    auto_seq_ch_sel = 0x12
+  };
+
+  void set_analog_channel(hal::byte p_channel);
+  void reset();
+
+  hal::i2c& m_bus;
+  hal::byte m_i2c_address;
+  hal::byte m_channel = 0x08;  // stores selected channel to reduce i2c requests
+  hal::byte m_object_created = 0x00;  // tracks adapter channel reservations
```

Comment:
I'd recommend using `std::bitset` over a hal::btye. They'll take up the same space and it has easy apis for setting and clearing bits.

Reply:
made this change but then found std::bitset<8> uses 4 bytes rather than 1 based on my linter and what I could find online, if Libhal's compiler is more efficient then I have the changes stashed and ready to push.

Reply:
Good catch. They must be using a uint32 underneath. You can stick with the uint8t you have.